### PR TITLE
feat:完成RNN情感分析

### PR DESCRIPTION
**Pull request by yansirgood**

### CODE REVIEW

**Comment by Besthope-Official** [comment]
reset 下提交历史，不要把缓存文件上传到 git 仓库里（以及重复提交之前的一个notebook）! 再提交代码之前, 先同步下源仓库的代码

```bash
git pull upstream main
```

把分支树整理成下面这样

image

然后改一些代码, 例如把代码中出现的绝对路径都改成相对路径, 你可以把数据集放到 /RNN/data 里面:

config.py:

```py
TRAINING_DIR = r"data/train.tsv"
TESTING_DIR = r"data/test.tsv"
```

index_view,py:

```py
submission_df = pd.read_csv(r'data/sampleSubmission.csv')
prediction_df = pd.read_csv('SA_predict.csv')
```

train.py:

```py
train_set = NameDataset(TRAINING_DIR, train=True)
val_set = NameDataset(TRAINING_DIR, val=True)
train_loader = DataLoader(train_set, batch_size=BATCH_SIZE, shuffle=True)
val_loader = DataLoader(val_set, batch_size=BATCH_SIZE, shuffle=False)
```

建议把 train_set, val_set, train_loader, val_loader 放到 train.py 下初始化, 你这里 import 了 NameDataset, 在 dataset 里又引用了 config, 循环引用导致你又在dataset有一个device

**Comment by Besthope-Official** [comment]
把 return_loss_list 删掉, 画 loss 曲线就在 train 里面

```py
if __name__ == '__main__':
    classifier = RNNClassifier(NUM_CHARS, HIDDEN_SIZE, NUM_CLASS, droup_out,NUM_LAYERS).to(device)
    criterion = torch.nn.CrossEntropyLoss()
    optimizer = torch.optim.Adam(classifier.parameters(), lr=0.001)
    loss_list=[]
    patience_counter=0
    for epoch in range(1, NUM_EPOCHS + 1):
        train_loss, val_loss = train_and_validate(classifier, criterion, optimizer, train_loader, val_loader, device)
        print(f'Epoch {epoch}, Train Loss: {train_loss}, Val Loss: {val_loss}')
        loss_list.append(val_loss)
        if val_loss <= min(loss_list):
            torch.save(classifier, 'sentimentAnalyst.pkl')
            print('Save Model!')
        else:
            patience_counter+=1
            if patience_counter >=patience:
                print('Triggered Early Stop')
                break
    epoch = [epoch + 1 for epoch in range(len(loss_list))]
    plt.plot(epoch, loss_list)
    plt.xlabel('Epoch')
    plt.ylabel('Loss')
    plt.grid()
    plt.show()
```

test.py

```py
if __name__ == '__main__':
    testModel()
```

**Comment by Besthope-Official** [comment]
最后的提交只要包含下面的源代码. README 里面请指出使用说明.

image

其他部分就不用改了, 模型没啥问题, 不用刷指标, Good Job!

**Comment by Besthope-Official** [comment]
调整到另一个分支合并
